text/0000-std-function-macro.md

```diff
@@ -0,0 +1,153 @@
+- Feature Name: Function Name Macro
+- Start Date: (2016-08-19)
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+`function_name` is a macro, which expands to the fully qualified function name
+from which it was invoked. It works similar to `std::file`, `std::line` or
+`std::module_path` and expands to an expression of type `&'static str`.
+
+# Motivation
+[motivation]: #motivation
+
+There are currently macros for getting the source file name, `std::file`, and
+the line number, `std::line`. To complete this pair, a macro to for getting
+the current function name would be useful. All three macros are useful for
+e.g. generating log output with more context about where something was
+happening. The file and line is good for being able to look up the exact
+location, however the fully qualified function name gives additional context
+without even having to look into the source code.
+
+# Detailed design
+[design]: #detailed-design
+
+This feature would have an equivalent design and implementation to
+`std::file`. It would return a fully qualified function name but the actual
+string returned is implementation defined. The string is meant to be used by
+the user to get an idea about the context of the code in e.g. debug output,
+and should be meaningful for a human, but is not meant to be analyzed by a
+program.
+
+For example the following scheme could be used:
+
+ - Plain functions use their name, e.g. `some_function` for
+
+````
+    fn some_function(arg: SomeThing) {
+    ...
+    }
+````
+
+ - Generic functions use their name, followed by the generic arguments, e.g.
+   `some_function<T, U>` for
```

Comment:
> If you see some_function:: you would know that it's this generic implementation, if you see some_function:: you know that it's a special implementation for exactly those types

Are you planning to output the whole set of bounds for each parameter then? If you want to avoid monomorphisations, then you won’t be able to uniquely identify the functions either (even more true in presence of specialisation).